server/src/main/java/org/opensearch/search/aggregations/bucket/FilterRewriteHelper.java

```diff
@@ -148,29 +155,29 @@ private static Weight[] createFilterForAggregations(
             // Below rounding is needed as the interval could return in
             // non-rounded values for something like calendar month
             roundedLow = preparedRounding.round(roundedLow + interval);
-            if (prevRounded == roundedLow) break;
+            if (prevRounded == roundedLow) break; // prevents getting into an infinite loop
```

Comment:
Can we use the size information in this loop to only generate the buckets equal to size? Also, we can short circuit the logic if min(bucketCount, size) is more than MAX_NUM_FILTER_BUCKETS

Reply:
in composite aggregation, size is the number of buckets with non-zero doc counts. In this stage (create filters), we don't know which filter may return 0, so we can only create as many as we can.

Later, at try filters stage, we loop over all the filters and only save the buckets with >0 count to the results, and we can terminate early there once we have enough size of non-zero buckets